trin-execution/src/evm/block_executor.rs

```diff
@@ -0,0 +1,285 @@
+    }
+
+    pub fn process_dao_fork(&mut self) -> anyhow::Result<()> {
+        // drain balances from DAO hardfork accounts
+        let drained_balances = self.evm.db_mut().drain_balances(DAO_HARDKFORK_ACCOUNTS)?;
+        let drained_balance_sum: u128 = drained_balances.iter().sum();
+
+        // transfer drained balance to beneficiary
+        self.increment_balances([(DAO_HARDFORK_BENEFICIARY, drained_balance_sum)].into_iter())?;
+
+        Ok(())
+    }
+
+    pub fn commit_bundle(&mut self) -> anyhow::Result<RootWithTrieDiff> {
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["commit_bundle"]);
+        self.evm
+            .db_mut()
+            .merge_transitions(BundleRetention::PlainState);
+        let state_bundle = self.evm.db_mut().take_bundle();
+        self.evm.db_mut().database.commit_bundle(state_bundle)?;
+        stop_timer(timer);
+
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["get_root_with_trie_diff"]);
+        let RootWithTrieDiff {
+            root,
+            trie_diff: changed_nodes,
+        } = self
+            .evm
+            .db_mut()
+            .database
+            .trie
+            .lock()
+            .root_hash_with_changed_nodes()?;
+
+        stop_timer(timer);
+
+        Ok(RootWithTrieDiff {
+            root,
+            trie_diff: changed_nodes,
+        })
+    }
+
+    pub fn execute_block(&mut self, block: &ProcessedBlock) -> anyhow::Result<()> {
+        info!("State EVM processing block {}", block.header.number);
+
+        let execute_block_timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["execute_block"]);
+        self.set_evm_environment_from_block(block);
+
+        // execute transactions
+        let cumulative_transaction_timer =
+            start_timer_vec(&BLOCK_PROCESSING_TIMES, &["cumulative_transaction"]);
+        let mut block_gas_used = 0;
+        for transaction in block.transactions.iter() {
+            let transaction_timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["transaction"]);
+            let evm_result = self.execute_transaction(transaction)?;
+            block_gas_used += evm_result.result.gas_used();
+            self.commit(evm_result.state);
+            stop_timer(transaction_timer);
+        }
+        stop_timer(cumulative_transaction_timer);
+
+        ensure!(
+            block_gas_used == block.header.gas_used.to::<u64>(),
+            "Block gas used mismatch at {} != {}",
+            block_gas_used,
+            block.header.gas_used
+        );
+        self.cumulative_gas_used += block_gas_used;
+
+        // update beneficiary
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["update_beneficiary"]);
+        let _ = self.increment_balances(get_block_reward(block));
+
+        // check if dao fork, if it is drain accounts and transfer it to beneficiary
+        if block.header.number == get_spec_block_number(SpecId::DAO_FORK) {
+            self.process_dao_fork()?;
```

Comment:
This should be done after the block is executed.

- block rewards
- beacon withdrawals
- dao fork draining

are all post block executed state changes.

The order you do this does matter, as balance differences can affect block execution

Reply:
I was just talking about DAO_FORK changes (I know that rewards and withdrawals must go at the end).

Since we are talking about one specific block, I can imagine that order in that specific block might not matter (regardless of which one is consider correct).

My feeling is that all fork related changes (both to state and evm execution) happen at the start of the block (before transactions are executed), therefore my question.
For example, what would happen if one of the malicious DAO wallets created transaction to transfer eth to "clean" wallet?! Current implementation wouldn't prevent that, but if this was executed at the start of the block, it would.

Ultimately, it doesn't matter (if we end up with the correct state), but logically it makes sense that dao_fork logic happens at the start of the block (at least for me).

Reply:
> For example, what would happen if one of the malicious DAO wallets created transaction to transfer eth to "clean" wallet?! Current implementation wouldn't prevent that, but if this was executed at the start of the block, it would.

The dao fork happened 8 years ago. The code to drain all the accounts was most likely a lot more complex then this to track if a transfer was to happen from the origin address.

We don't have to prevent anything so I don't really understand where this hypothetical is coming from. After the fork happened we would no longer need the complex code to properly prevent a transfer, we would only need to drain the accounts and transfer the funds at the right block.

I don't think discussing an 8 year old hypothetical is productive. This time should be spent working on shipping latest state data on the State network


> My feeling is that all fork related changes (both to state and evm execution) happen at the start of the block (before transactions are executed), therefore my question.

My feeling is all the balance_increment code should be together, because that is what this is.

Reply:
> My feeling is that all fork related changes (both to state and evm execution) happen at the start of the block (before transactions are executed)

The 
- change in block reward are fork related changes (certain forks changed what the block reward number was)
- withdrawals are fork related changes (these were activated on shanghai)
- the dao fork draining and moving the funds are fork related change

So I do not see why there is a distinction for the dao fork, when all 3 of those relate to forks.

The commonality is they all use `increment_balances``

Reply:
My meaning of "fork related changes" are the changes either to the evm (or evm state) that happen **only once**, at the fork block. Some forks enable/disable/modify certain opt-codes, modify gas calculations, allow certain behaviors (withdrawals), etc. In my mental model, these changes happen at the start of the fork block (and only then).

Dao fork is not much different in that sense, at least to me (because it happens only once).

The other two (block rewards, withdrawals) happen at the end of the block, which is correct because it's according to spec. They are fork related because certain forks enable/modify them, but once they are active, they happen on **every** block afterwards.

Reply:
And I agree with you that there is no point in discussing this further (as it's just one block that happened 8 years ago).

Feel free to leave it as it is (I just wanted to clarify my mental model).